Доработана ф-ность по  ТЗ спринта 6

Ycseeasy opened the pull request.

Ycseeasy committed:
Добавил реализацию 6 ТЗ:
Добавил дополнительный класс для узлов NODE

Ycseeasy committed:
Добавил реализацию 6 ТЗ:
Добавил дополнительные тестирования.

Ycseeasy committed:
Добавил реализацию 6 ТЗ:
Удалил ненужный сеттер списка сабтасок

Ycseeasy committed:
Добавил реализацию 6 ТЗ:
Добавил новый метод remove, для ТЗ

Ycseeasy committed:
Добавил реализацию 6 ТЗ:
Добавил реализацию двухсвязного списка по ТЗ + метод удаления тасок

Ycseeasy committed:
Добавил реализацию 6 ТЗ:
Обновил методы удаления тасок

annaalkh commented on src/taskManagers/Node.java:
```diff
@@ -0,0 +1,17 @@
+
+import tasks.Task;
+
+public class Node<Task> {
```
Использование дженериков само по себе верное, но здесь не имеет особого смысла - класс Node определен и используется внутри InMemoryHistoryManager, который сам по себе работает только с задачами - поэтому ни для каких других данных Node мы явно не будем использовать. А для хранения любых типов задач достаточно, чтобы Node работал с типом Task - т.к. остальные задачи наследники Task

annaalkh commented on src/taskManagers/InMemoryHistoryManager.java:
```diff
@@ -2,26 +2,85 @@
+        if (head == null)
+            head = temp;
+        else
+            tail.next = temp;
```
Лучше всегда использовать фигурные скобки с if-else - такой синтаксис может вызвать ошибки (например, если понадобится добавить еще одно действие в else, можно забыть добавить скобки, и программа будет работать некорректно)

annaalkh commented on src/taskManagers/InMemoryHistoryManager.java:
```diff
@@ -2,26 +2,85 @@
+            if (removeNode == tail)
+                removeLast();
+            else
+                removeNode.next.prev = removeNode.prev;
```
Есть вопрос, будет ли этот алгоритм корректно работать, если удаляется единственная нода в списке - это нужно проверить, а еще лучшие - добавить тест для этой ситуации

annaalkh commented on src/taskManagers/InMemoryHistoryManager.java:
```diff
@@ -2,26 +2,85 @@
+        return temp;
+    }
+
+    public void removeFirst() {
```
Методы обозначены как public - значит, их может вызывать код из других классов. Но, во первых это не совсем соответствует условию задачи, а главное - если вызвать их у пустого списка, метод упадет с NullPointerException (а другие классы не могут знать, что так делать не нужно). Поэтому стоит сделать методы removeFirst и removeLast приватными

annaalkh commented on src/taskManagers/InMemoryHistoryManager.java:
```diff
@@ -2,26 +2,85 @@
-            taskHistory.add(taskForHistory);
-        } else {
-            taskHistory.add(taskForHistory);
+        if (!hashId.containsKey(task.getId())) {
```
По условию, если задача уже есть в истории, старое вхождение должно быть удалено, а новое добавлено в конец списка - сейчас такой сценарий не реализован

annaalkh commented on test/ProjectTest.java:
```diff
@@ -26,79 +27,98 @@ public void taskEquals() {
-        assertNotEquals(history.get(0).getName(),history.get(1).getName());
-        assertNotEquals(history.get(0).getDescription(),history.get(1).getDescription());
-        assertNotEquals(history.get(0).getTaskStatus(),history.get(1).getTaskStatus());
+        assertNotEquals(history.get(0), history.get(1));
```
Здесь было бы лучше сравнить задачи с ожидаемыми, а не между собой - то, что в истории находятся две разные задачи, никак не гарантирует, что это верные задачи

annaalkh commented on test/ProjectTest.java:
```diff
@@ -26,79 +27,98 @@ public void taskEquals() {
-        assertNotEquals(history.get(0).getDescription(),history.get(1).getDescription());
-        assertNotEquals(history.get(0).getTaskStatus(),history.get(1).getTaskStatus());
+        assertNotEquals(history.get(0), history.get(1));
+        assertEquals(history.size(), 2);
```
Эту проверку лучше сделать первой - если по какой-то причине в истории окажется меньше задач, тест должен выдать ошибку сравнения, а не упасть с исключением, как будет сейчас

annaalkh reviewed (commented):
Илья, привет! В целом реализация верная, но есть отдельные моменты, которые нужно доработать (самое главное - логика добавления новой задачи в условию, требуемый сценарий реализован не полностью). Также рекомендую не использовать if-else без фигурных скобок (писала в комментарии, но здесь хочу подчеркнуть - т.к. не раз сталкивалась с проблемами в проектах из-за такого синтаксиса)

Ycseeasy committed:
Добавил реализацию 6 ТЗ:
Обновил методы удаления тасок

Ycseeasy committed:
Работа с ревью(1):
Поставил раньше проверку по размеру списка и добавил проверку на совпадении с ожидаемым результатом

Ycseeasy committed:
Работа с ревью(1):
Отказался от джереника

Ycseeasy committed:
Работа с ревью(1):
Попраил ошибки связанные с реализацией ТЗ (добавление/удаление нодов)

annaalkh reviewed (approved):
Отлично, работа принята!
